### PR TITLE
Add compatibility notice to sidebar when editing Cart and Checkout blocks (and their inner blocks)

### LINKED ISSUE #6851: Editor: Add an alert notice about possible incompatibilities to the Cart and Checkout Blocks
## Describe the bug

The Cart and Checkout Blocks's editor sidebar will include an alert that warns about possible incompatibilities.

Figma file bA1eV1oHbHs3sEEVnXbj4h-fi-383%3A10581
Context: pdFofs-b7-p2

The alert will:
- be placed under the Name & description panel, before the Block's settings panels
- the link will point to the this [Incompatible Extensions section](https://woocommerce.com/document/cart-checkout-blocks-support-status/#section-7)
- will appear when selecting the parent Cart and Checkout Blocks as well as their inner blocks

Behaviour:
- the alert is replacing the existing [pop-up alert ](https://user-images.githubusercontent.com/1628454/182843881-7847393d-8f1e-4373-93b8-5d75e1f26bfc.png)
- the alert will appear even for users that have previously dismissed the current [pop-up alert ](https://user-images.githubusercontent.com/1628454/182843881-7847393d-8f1e-4373-93b8-5d75e1f26bfc.png)
- once it’s dismissed it wouldn’t appear again

## Expected behavior

0. As an user I create a new page and insert the Cart or the Checkout Blocks for the first time
1. I no longer see the [pop-up alert ](https://user-images.githubusercontent.com/1628454/182843881-7847393d-8f1e-4373-93b8-5d75e1f26bfc.png)
2. When I select the Block or one of it's inner blocks I notice the alert, under the name and description section in the Block's sidebar
3. I dismiss the alert and navigate away from the page
4. I come back to the page and I no longer see the alert

## Additional context

The new alert can use a dismissal mechanism similar to the one available for the current [pop-up alert](https://user-images.githubusercontent.com/1628454/182843881-7847393d-8f1e-4373-93b8-5d75e1f26bfc.png). Search for `wc-blocks_dismissed_compatibility_notices` 

The Mini Cart will still use  `wc-blocks_dismissed_compatibility_notices` so we won't remove it as part of this effort yet.

### PR DESCRIPTION
**Pull request by opr**
This PR adds a notice to the inspector when editing Cart and Checkout blocks, and their inner blocks. It also adds functionality to dismiss the notice independently on Cart and Checkout blocks.

I reused the `useCompatibilityNotice` functionality we have, and used the block names. We don't need to make new ones since `cart` and `checkout` were currently being used to track the old pop-up compatibility notice. This won't affect mini cart as that could use its own key to track its own compatibility notice status.

Fixes #6851

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

image

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

1. Before checking this PR out, build `trunk`, clear local storage and visit the Cart/Checkout blocks. Dismiss the compatibility notices on each page. Then check out this PR.
2. Go to the editor, insert the Checkout block.
3. Select it, verify you see the notice. Click through each and every inner block and ensure the notice displays for it.
4. Dismiss the notice, click back through the inner blocks and ensure the notice is gone. Reload the page, ensure the notice is still gone.
5. Go to the editor, insert the Cart block.
6. Select it, verify you see the notice. Click through each and every inner block and ensure the notice displays for it.
7. Dismiss the notice, click back through the inner blocks and ensure the notice is gone. Reload the page, ensure the notice is still gone.
8. Clear local storage again.
9. Go back to the Cart and Checkout blocks and ensure the notice is back.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [ ] WooCommerce Core
* [x] Feature plugin
* [ ] Experimental

### Changelog

> Add notice to Cart and Checkout blocks' inspector controls which links to the list of compatible plugins.

### CODE REVIEW

File: assets/js/blocks/cart/edit.js [hunk 1]
```diff
@@ -38,6 +38,7 @@ import {
 	useForcedLayout,
 } from '../cart-checkout-shared';
 import { CartBlockContext } from './context';
+import { CartSidebarCompatibilityNotice } from '@woocommerce/editor-components/sidebar-compatibility-notice';
```
Comment (tarhi-saad) [inline assets/js/blocks/cart/edit.js]: We should move this new notice component to the `External dependencies` list. And remove the old one in `line 15`. 🙌

File: assets/js/blocks/cart/edit.js [hunk 2]
```diff
@@ -111,6 +112,7 @@ const BlockSettings = ( { attributes, setAttributes } ) => {
 				/>
 			</PanelBody>
 			<CartCheckoutFeedbackPrompt />
+			<CartSidebarCompatibilityNotice />
```
Comment (tarhi-saad) [inline assets/js/blocks/cart/edit.js]: We should place the alert under the `Name & description` panel before the Block's settings panels (i.e., Let's move it to `line 70`).

File: assets/js/blocks/cart/inner-blocks/cart-order-summary-taxes/edit.tsx [hunk 3]
```diff
@@ -63,6 +64,7 @@ export const Edit = ( {
 							/>
 						</PanelBody>
 					) }
+				<CartSidebarCompatibilityNotice />
```
Comment (tarhi-saad) [inline assets/js/blocks/cart/inner-blocks/cart-order-summary-taxes/edit.tsx]: Same here! We should place the alert under the `Name & description` panel before the Block's settings panels (i.e., Let's move it to `line 39`).

File: assets/js/blocks/cart/inner-blocks/cart-order-summary-shipping/edit.tsx [hunk 4]
```diff
@@ -59,6 +60,7 @@ export const Edit = ( {
 						/>
 					</PanelBody>
 				) }
+				<CartSidebarCompatibilityNotice />
```
Comment (tarhi-saad) [inline assets/js/blocks/cart/inner-blocks/cart-order-summary-shipping/edit.tsx]: Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 37).

File: assets/js/blocks/cart/inner-blocks/proceed-to-checkout-block/edit.tsx [hunk 5]
```diff
@@ -59,6 +62,7 @@ export const Edit = ( {
 						} }
 					/>
 				) }
+				<CartSidebarCompatibilityNotice />
```
Comment (tarhi-saad) [inline assets/js/blocks/cart/inner-blocks/proceed-to-checkout-block/edit.tsx]: Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 45).

File: assets/js/blocks/checkout/edit.tsx [hunk 6]
```diff
@@ -29,6 +29,8 @@ import { CHECKOUT_PAGE_ID } from '@woocommerce/block-settings';
 import { createInterpolateElement } from '@wordpress/element';
 import { getAdminLink } from '@woocommerce/settings';
 import { CartCheckoutCompatibilityNotice } from '@woocommerce/editor-components/compatibility-notices';
+import { CheckoutSidebarCompatibilityNotice } from '@woocommerce/editor-components/sidebar-compatibility-notice';
+import {} from '@woocommerce/editor-components/compatibility-notices';
```
Comment (tarhi-saad) [inline assets/js/blocks/checkout/edit.tsx]: We should remove the imports at `line 33` & `line 31`!

File: assets/js/blocks/checkout/inner-blocks/checkout-order-summary-taxes/edit.tsx [hunk 7]
```diff
@@ -63,6 +64,7 @@ export const Edit = ( {
 							/>
 						</PanelBody>
 					) }
+				<CheckoutSidebarCompatibilityNotice />
```
Comment (tarhi-saad) [inline assets/js/blocks/checkout/inner-blocks/checkout-order-summary-taxes/edit.tsx]: Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 39).

File: assets/js/blocks/checkout/inner-blocks/checkout-payment-block/edit.tsx [hunk 8]
```diff
@@ -85,6 +86,7 @@ export const Edit = ( {
 						</ExternalLink>
 					</PanelBody>
 				) }
+				<CheckoutSidebarCompatibilityNotice />
```
Comment (tarhi-saad) [inline assets/js/blocks/checkout/inner-blocks/checkout-payment-block/edit.tsx]: Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 56).

File: assets/js/blocks/checkout/inner-blocks/checkout-shipping-methods-block/edit.tsx [hunk 9]
```diff
@@ -119,6 +120,7 @@ export const Edit = ( {
 						</ExternalLink>
 					</PanelBody>
 				) }
+				<CheckoutSidebarCompatibilityNotice />
```
Comment (tarhi-saad) [inline assets/js/blocks/checkout/inner-blocks/checkout-shipping-methods-block/edit.tsx]: Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 60).

File: assets/js/blocks/checkout/inner-blocks/checkout-terms-block/edit.tsx [hunk 10]
```diff
@@ -55,6 +56,7 @@ export const Edit = ( {
 						}
 					/>
 				</PanelBody>
+				<CheckoutSidebarCompatibilityNotice />
```
Comment (tarhi-saad) [inline assets/js/blocks/checkout/inner-blocks/checkout-terms-block/edit.tsx]: Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 40).

File: assets/js/blocks/checkout/edit.tsx [hunk 11]
```diff
@@ -109,6 +111,7 @@ const BlockSettings = ( {
 				/>
 			</PanelBody>
 			<CartCheckoutFeedbackPrompt />
+			<CheckoutSidebarCompatibilityNotice />
```
Comment (tarhi-saad) [inline assets/js/blocks/checkout/edit.tsx]: Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 69).

File: assets/js/blocks/checkout/edit.tsx [hunk 12]
```diff
@@ -166,6 +169,7 @@ export const Edit = ( {
 					}
 				/>
 			</PanelBody>
+			<CheckoutSidebarCompatibilityNotice />
```
Comment (tarhi-saad) [inline assets/js/blocks/checkout/edit.tsx]: Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 153).

File: assets/js/blocks/checkout/edit.tsx [hunk 13]
```diff
@@ -225,6 +229,7 @@ export const Edit = ( {
 					/>
 				) }
 			</PanelBody>
+			<CheckoutSidebarCompatibilityNotice />
```
Comment (tarhi-saad) [inline assets/js/blocks/checkout/edit.tsx]: Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 178).

File: assets/js/editor-components/sidebar-compatibility-notice/style.scss [hunk 14]
```diff
@@ -0,0 +1,13 @@
+.wc-blocks-sidebar-compatibility-notice {
+	border-top: 1px solid $gray-200;
+}
+
+.wc-blocks-sidebar-compatibility-notice__notice.components-notice.is-dismissible {
+	margin: 0;
+	padding: $gap;
+	padding-right: $gap-small;
```
Comment (tarhi-saad) [inline assets/js/editor-components/sidebar-compatibility-notice/style.scss]: We have more `padding` to the right which produces an uneven spacing.

image

File: assets/js/editor-components/sidebar-compatibility-notice/style.scss [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+
+	.components-notice__content {
+		margin-right: 0;
```
Comment (tarhi-saad) [inline assets/js/editor-components/sidebar-compatibility-notice/style.scss]: This style isn't applied on my end!

image

File: assets/js/editor-components/sidebar-compatibility-notice/style.scss [hunk 16]
```diff
@@ -0,0 +1,13 @@
+.wc-blocks-sidebar-compatibility-notice {
+	border-top: 1px solid $gray-200;
```
Comment (tarhi-saad) [inline assets/js/editor-components/sidebar-compatibility-notice/style.scss]: There is no `border-top` from my side!

image

File: assets/js/editor-components/sidebar-compatibility-notice/index.tsx [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+	const noticeText = createInterpolateElement(
+		__(
+			'The Cart & Checkout Blocks are a beta feature to optimize for faster checkout. To make sure this feature is right for your store, <a>review the list of compatible extensions</a>.',
+			'woo-gutenberg-products-block'
+		),
+		{
+			a: (
+				// Suppress the warning as this <a> will be interpolated into the string with content.
+				// eslint-disable-next-line jsx-a11y/anchor-has-content
+				<ExternalLink href="https://woocommerce.com/document/cart-checkout-blocks-support-status/#section-7" />
```
Comment (tarhi-saad) [inline assets/js/editor-components/sidebar-compatibility-notice/index.tsx]: We should redirect to the list of `Compatible Extensions` (i.e., this link instead: https://woocommerce.com/document/cart-checkout-blocks-support-status/#section-3). In the issue description that @ralucaStan wrote #6851, we are mentioning the list of `Incompatible Extensions` instead. But, on the Figma file and on the context post pdFofs-b7-p2, we want to redirect to the `Compatible Extensions`.

**Review by tarhi-saad — CHANGES_REQUESTED** [review]
Thank you, @opr, for working on this PR! 🙌  I've left some comments below! 👍 I apologize for leaving a lot of comments regarding the same issue: 

```
The alert will:

be placed under the Name & description panel, before the Block's settings panels
``` 
I just wanted to make it easy to find all the files and lines that should be updated! 🙌

**Comment by opr** [comment]
Hey @tarhi-saad thanks a lot for the detailed review! I actually moved things about a bit more as sometimes it wasn't possible to position the notice at the top due to other controls being used.

I've added a `includeCompatibilityNotice` prop to the `addressFieldControls` to allow us to turn the notice off if required. This was needed for the `checkout fields` block since it was showing two notices. Trying to get it to only show one affected the address field blocks, and the contact information block.

I also changed the stylesheet to be called `editor.scss` so it'll be included in our editor stylesheet.

**Pull request closed without merging**